kernel/src/mem.rs

```diff
@@ -1,202 +1,281 @@
+    /// address `0x0`.
+    fn ptr(&self) -> *const u8;
 
-/// Base type for an AppSlice that holds the raw pointer to the memory region
-/// the app shared with the kernel.
-pub(crate) struct AppPtr<L, T> {
-    ptr: NonNull<T>,
-    process: AppId,
-    _phantom: PhantomData<L>,
+    /// Applies a function to the (read only) slice reference pointed
+    /// to by the AppSlice.
+    ///
+    /// If the process is no longer alive and the memory has been
+    /// reclaimed, this method must return the default value.
+    ///
+    /// # Default AppSlice
+    ///
+    /// A default instance of an AppSlice must return the passed
+    /// default value without executing the closure.
+    fn map_or<F, R>(&self, default: R, fun: F) -> R
+    where
+        F: FnOnce(&[u8]) -> R;
 }
 
-impl<L, T> AppPtr<L, T> {
-    /// Safety: Trusts that `ptr` points to userspace memory in `appid`
-    unsafe fn new(ptr: NonNull<T>, appid: AppId) -> AppPtr<L, T> {
-        AppPtr {
-            ptr: ptr,
-            process: appid,
-            _phantom: PhantomData,
-        }
-    }
+/// A readable and writable region of userspace memory.
+///
+/// This trait can be used to gain read-write access to memory regions
+/// wrapped in an AppSlice type.
+///
+/// This is a supertrait of [`Read`], which features further methods.
+pub trait ReadWrite: Read {
+    /// Mutable pointer to the userspace memory region
+    ///
+    /// If the length of the initially shared memory region
+    /// (irrespective of the return value of [`len`](Read::len)) is 0,
+    /// this function must return a pointer to address `0x0`. This is
+    /// because processes allow buffers with length 0 to reclaim
+    /// shared memory with the kernel and are allowed to specify _any_
+    /// address, even if it is not contained within their address
+    /// space. These _dummy addresses_ should not be leaked to outside
+    /// code.
+    fn mut_ptr(&self) -> *mut u8;
+
+    /// Applies a function to the mutable slice reference pointed to
+    /// by the AppSlice.
+    ///
+    /// If the process is no longer alive and the memory has been
+    /// reclaimed, this method must return the default value.
+    ///
+    /// # Default AppSlice
+    ///
+    /// A default instance of an AppSlice must return the passed
+    /// default value without executing the closure.
+    fn mut_map_or<F, R>(&self, default: R, fun: F) -> R
+    where
+        F: FnOnce(&mut [u8]) -> R;
 }
 
-/// Buffer of memory shared from an app to the kernel.
-///
-/// This is the type created after an app calls the `allow` syscall.
-pub struct AppSlice<L, T> {
-    ptr: AppPtr<L, T>,
+/// Read-writable memory region of a process, shared with the kernel
+pub struct ReadWriteAppSlice {
+    ptr: *mut u8,
     len: usize,
+
+    // TODO: For improved efficiency this should use a dummy instance
+    // of an AppId when constructed with Default::default(), as the
+    // Option will allocate another full word.
```

Comment:
This seems like a good use case for `NonZeroU32`: https://doc.rust-lang.org/core/num/struct.NonZeroU32.html . We would just have to promise that no apps can have AppId == 0

Reply:
Actually, this TODO comment is a lie: `mem::size_of::<AppId>() == mem::size_of::<Option<AppId>>()`, because the first field in `AppId` is a reference to the kernel which can never be zero. Removed it accordingly, sorry for bothering you with this!